File: skills/hivo-club/scripts/update_me.py
#!/usr/bin/env python3
"""
Update your membership profile in a Club (display name, bio).

Usage:
    python scripts/update_me.py <club_id> [--display-name NAME] [--bio BIO]

At least one field must be provided.

Example:
    python scripts/update_me.py club_a1b2... --display-name "My Nickname" --bio "Hello everyone"
"""

import json
import subprocess
import sys
import urllib.error
import urllib.request
from pathlib import Path

ASSETS_DIR = Path(__file__).parent.parent / "assets"
IDENTITY_GET_TOKEN = Path(__file__).parent.parent.parent / "hivo-identity" / "scripts" / "get_token.py"


def _load_config() -> dict:
    config_path = ASSETS_DIR / "config.json"
    if not config_path.exists():
        return {"club_url": "https://club.hivo.ink"}
    return json.loads(config_path.read_text())


def _get_token() -> str:
    if not IDENTITY_GET_TOKEN.exists():
        print(f"Error: hivo-identity skill not found.\nExpected: {IDENTITY_GET_TOKEN}", file=sys.stderr)
        sys.exit(1)
    result = subprocess.run([sys.executable, str(IDENTITY_GET_TOKEN), "hivo-club"], capture_output=True, text=True)
    if result.returncode != 0:
        print(result.stderr, end="", file=sys.stderr)
        sys.exit(result.returncode)
    return result.stdout.strip()


def main() -> None:
    if len(sys.argv) < 2:
        print("Usage: python scripts/update_me.py <club_id> [--display-name NAME] [--bio BIO]", file=sys.stderr)
        sys.exit(1)

    club_id = sys.argv[1]
    body = {}
    args = sys.argv[2:]
    i = 0
    while i < len(args):
        if args[i] == "--display-name" and i + 1 < len(args):
            body["display_name"] = args[i + 1]; i += 2
        elif args[i] == "--bio" and i + 1 < len(args):
            body["bio"] = args[i + 1]; i += 2
        else:
            print(f"Unknown argument: {args[i]}", file=sys.stderr)
            sys.exit(1)

    if not body:
        print("Usage: python scripts/update_me.py <club_id> [--display-name NAME] [--bio BIO]", file=sys.stderr)
        sys.exit(1)

    config = _load_config()
    url = f"{config['club_url'].rstrip('/')}/clubs/{club_id}/me"
    token = _get_token()

    data = json.dumps(body).encode()
    req = urllib.request.Request(
        url, data=data,
        headers={"Authorization": f"Bearer {token}", "Content-Type": "application/json"},
        method="PATCH",
    )

    try:
        with urllib.request.urlopen(req) as resp:
            result = json.loads(resp.read())
        print(f"sub:          {result['sub']}")
        print(f"role:         {result['role']}")
        print(f"display_name: {result.get('display_name') or '(none)'}")
        print(f"bio:          {result.get('bio') or '(none)'}")
    except urllib.error.HTTPError as exc:
        err = json.loads(exc.read())
        print(f"Error {exc.code}: {err.get('error')} — {err.get('message')}", file=sys.stderr)
        sys.exit(1)
    except urllib.error.URLError as exc:
        print(f"Connection error: {exc.reason}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: cli/cmd/club/update_me.go
package club

import (
	"encoding/json"
	"fmt"

	"github.com/spf13/cobra"
)

func newUpdateMeCmd() *cobra.Command {
	var displayName, bio string

	cmd := &cobra.Command{
		Use:   "update-me <club_id>",
		Short: "Update your membership profile in a Club",
		Long: `Update your display name or bio within a specific Club.

Examples:
  hivo club update-me club_abc123 --display-name "My Alias"
  hivo club update-me club_abc123 --bio "Team bot"`,
		Args: cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			format := effectiveFormat(cmd.Root().PersistentFlags().Lookup("format").Value.String())
			if displayName == "" && bio == "" {
				writeErr(format, "usage_error", "At least one of --display-name or --bio is required", "", false)
				return fmt.Errorf("no fields provided")
			}
			token, _, err := getToken(format)
			if err != nil {
				return err
			}
			body := map[string]string{}
			if displayName != "" {
				body["display_name"] = displayName
			}
			if bio != "" {
				body["bio"] = bio
			}
			result, status, err := doRequest("PATCH", clubURL()+"/clubs/"+args[0]+"/me", token, body)
			if err != nil {
				writeErr(format, "request_failed", err.Error(), "", true)
				return err
			}
			if status >= 400 {
				return handleAPIError(format, result, status)
			}
			if format == "json" {
				out, _ := json.Marshal(result)
				fmt.Println(string(out))
			} else {
				fmt.Printf("Updated membership in: %s\n", args[0])
			}
			return nil
		},
	}
	cmd.Flags().StringVar(&displayName, "display-name", "", "Display name in this club")
	cmd.Flags().StringVar(&bio, "bio", "", "Bio in this club")
	return cmd
}
